### PR TITLE
data flow for audio repeat feature

### PR DESCRIPTION
**Pull request by muhajirdev**
### Summary

Status: functional, but I am still testing for corner cases
Code: Not well organized yet.

### CODE REVIEW

File: src/components/AudioPlayer/AudioRepeatManager/hooks.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  useEffect(() => {
+    if (!repeatSettings.from || !repeatSettings.to) return null;
+    repeatVerse.current.progress = 1;
+    repeatVerse.current.total = repeatSettings.repeatEachVerse;
+    return null;
+  }, [repeatSettings.from, repeatSettings.repeatEachVerse, repeatSettings.to]);
+
+  return repeatVerse;
+};
+
```
Comment (tifa2UP) [inline src/components/AudioPlayer/AudioRepeatManager/hooks.ts:32]: refactored to redux

File: src/components/AudioPlayer/AudioRepeatManager/hooks.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+// save the delayMultiplier data to the `ref`
+// and reset the data when new data is received from redux
+export const useDelayMultiplier = (repeatSettings) => {
+  const delayMultiplierBetweenVerse = useRef(0);
```
Comment (tifa2UP) [inline src/components/AudioPlayer/AudioRepeatManager/hooks.ts:36]: All this should be retrieved from redux and not managed on a local level.

File: src/components/AudioPlayer/AudioRepeatManager/utils.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
+import { triggerPauseAudio, triggerPlayAudio } from '../EventTriggers';
+
+// calculate the duration between timestampFrom and timestampTo * delayMultiplierBetweenVerse
+const getDelay = ({ delayMultiplierBetweenVerse, verseTiming }) => {
+  const { timestampTo, timestampFrom } = verseTiming;
+  return delayMultiplierBetweenVerse * (timestampTo - timestampFrom);
+};
+
+// based on given conditions, decide when to stop the audio and when to delay the audio
```
Comment (tifa2UP) [inline src/components/AudioPlayer/AudioRepeatManager/utils.ts:9]: Be specific on the conditions

File: src/components/AudioPlayer/AudioRepeatManager/utils.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  return {
+    shouldRepeatVerse,
+    shouldResetVerseProgress,
+    shouldRepeatRange,
+    shouldStopAudio,
+    shouldDelayAudio,
+  };
+};
+
+// get new time for audio player based on given data
```
Comment (tifa2UP) [inline src/components/AudioPlayer/AudioRepeatManager/utils.ts:62]: 1) Add better documentation

File: src/components/AudioPlayer/AudioRepeatManager/AudioRepeatManager.tsx [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+      verseRangeTo,
+    });
+
+    const newtime = getNewTime({
+      lastActiveVerseTiming,
+      shouldRepeatRange,
+      shouldRepeatVerse,
+      verseRangeFrom,
+    });
+    if (typeof newtime === 'number') triggerSetCurrentTime(newtime);
```
[the remark this answers is not in the document]
Comment (muhajirdev) [reply src/components/AudioPlayer/AudioRepeatManager/AudioRepeatManager.tsx:90]: image

I think isNaN is more loose

File: src/components/QuranReader/TranslationView/TranslationViewCell.tsx [hunk 6]
```diff
@@ -64,10 +64,7 @@ const TranslationViewCell = ({ verse }: TranslationViewCellProps) => {
           </div>
           <div className={styles.actionContainerRight}>
             <div className={styles.actionItem}>
-              <PlayVerseAudioButton
-                timestamp={verse.timestamps.timestampFrom}
-                verseKey={verse.verseKey}
-              />
+              <PlayVerseAudioButton verseKey={verse.verseKey} />
```
[the remark this answers is not in the document]
Comment (muhajirdev) [reply src/components/QuranReader/TranslationView/TranslationViewCell.tsx:67]: can I do this in the follow up PR? this PR is getting too big

File: src/components/AudioPlayer/RepeatAudioModal/RepeatAudioModal.tsx [hunk 7]
```diff
@@ -47,29 +54,54 @@ const RepeatAudioModal = ({
   const lastVerseKeyInThisChapter = comboboxVerseItems[comboboxVerseItems.length - 1].value;
 
   const [verseRepetition, setVerseRepetition] = useState({
-    repeatRange: 1,
+    repeatRange: 3,
     repeatEachVerse: 1,
     from: defaultSelectedVerse || firstVerseKeyInThisChapter,
-    to: lastVerseKeyInThisChapter,
+    to: defaultSelectedVerse || lastVerseKeyInThisChapter,
+    delayMultiplierBetweenVerse: 0,
```
[the remark this answers is not in the document]
Comment (muhajirdev) [reply src/components/AudioPlayer/RepeatAudioModal/RepeatAudioModal.tsx]: solved 1 and 2. 
3. will be in the follow up PR

File: src/components/AudioPlayer/AudioRepeatManager/utils.ts (hunk 4 above)
Comment (muhajirdev) [reply src/components/AudioPlayer/AudioRepeatManager/utils.ts]: deleted this file

File: src/components/AudioPlayer/AudioRepeatManager/utils.ts (hunk 3 above)
Comment (muhajirdev) [reply src/components/AudioPlayer/AudioRepeatManager/utils.ts]: deketed this file

File: src/components/AudioPlayer/AudioRepeatManager/hooks.ts (hunk 2 above)
Comment (muhajirdev) [reply src/components/AudioPlayer/AudioRepeatManager/hooks.ts]: deleted this file

File: src/components/AudioPlayer/AudioRepeatManager/hooks.ts (hunk 1 above)
Comment (muhajirdev) [reply src/components/AudioPlayer/AudioRepeatManager/hooks.ts]: deleted this file

File: src/components/AudioPlayer/AudioRepeatManager/AudioRepeatManager.tsx [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+    // 2) pause the audio when delayMultiplier is set
+    // 3) increment repeatEachVerse progress by 1
+    if (isVerseEnded && repeatProgress.repeatEachVerse < repeatSettings.repeatEachVerse) {
+      triggerSetCurrentTime(lastActiveVerseTiming.current.timestampFrom / 1000);
+      delayAudioWhenNeeded();
+      dispatch(setRepeatProgress({ repeatEachVerse: repeatProgress.repeatEachVerse + 1 }));
+      return null;
+    }
+
+    // When verse ended, repeatEachVerse progress === expected repeatEachProgress. reset the value
```
Comment (tifa2UP) [inline src/components/AudioPlayer/AudioRepeatManager/AudioRepeatManager.tsx:120]: update the comment to clarify the if statement

File: src/components/QuranReader/TranslationView/TranslationViewCell.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (tifa2UP) [reply src/components/QuranReader/TranslationView/TranslationViewCell.tsx:67]: yup

File: src/utils/verse.ts [hunk 9]
```diff
@@ -312,3 +313,15 @@ export const isLastVerseOfSurah = (chapterNumber: string, verseNumber: number):
...
+  const chapterData = getChapterData(chapterId);
+  return [
+    makeVerseKey(Number(chapterData.id), 1),
+    makeVerseKey(Number(chapterData.id), chapterData.versesCount),
+  ];
+};
+
+export const getVerseTimingByVerseKey = (verseKey: string, verseTimings: VerseTiming[]) => {
+  return verseTimings.find((verseTiming) => verseTiming.verseKey === verseKey);
+};
```
Comment (tifa2UP) [inline src/utils/verse.ts:327]: nit: sounds to me like an audio util

**Pull request merged**